lib/spack/spack/environment.py

```diff
@@ -261,132 +261,152 @@ def apply_modifications(self):
-                if separator in first_value and separator in second_value:
-                    return separator
-            return None
+                for arg in args:
+                    if separator in arg:
+                        return separator
 
-        for x in modified_variables:
-            current = this_environment[x]
-            modified = after_source_env[x]
-            sep = return_separator_if_any(current, modified)
-            if sep is None:
-                # We just need to set the variable to the new value
-                env.set(x, after_source_env[x])
+        # Add variables to env.
+        # Assume that variables with 'PATH' in the name or that contain
+        # separators like ':' or ';' are more likely to be paths
+        for x in new_variables:
+            sep = return_separator_if_any(env_after[x])
+            if sep or 'PATH' in x:
+                env.prepend_path(x, env_after[x])
             else:
-                current_list = current.split(sep)
-                modified_list = modified.split(sep)
+                # We just need to set the variable to the new value
+                env.set(x, env_after[x])
+
+        for x in unset_variables:
+            env.unset(x)
+
+        for x in modified_variables:
+            before = env_before[x]
+            after = env_after[x]
+            sep = return_separator_if_any(before, after)
+            if sep:
+                before_list = before.split(sep)
+                after_list = after.split(sep)
+
+                # Filter out empty strings
+                before_list = list(filter(None, before_list))
+                after_list = list(filter(None, after_list))
+
                 # Paths that have been removed
                 remove_list = [
-                    ii for ii in current_list if ii not in modified_list]
-                # Check that nothing has been added in the middle of vurrent
-                # list
+                    ii for ii in before_list if ii not in after_list]
+                # Check that nothing has been added in the middle of
+                # before_list
                 remaining_list = [
-                    ii for ii in current_list if ii in modified_list]
-                start = modified_list.index(remaining_list[0])
-                end = modified_list.index(remaining_list[-1])
-                search = sep.join(modified_list[start:end + 1])
+                    ii for ii in before_list if ii in after_list]
+                start = after_list.index(remaining_list[0])
+                end = after_list.index(remaining_list[-1])
+                search = sep.join(after_list[start:end + 1])
 
-                if search not in current:
+                if search not in before:
                     # We just need to set the variable to the new value
-                    env.set(x, after_source_env[x])
+                    env.set(x, env_after[x])
                     break
```

Comment:
Reading the code, I think this was a bug from my side. Here we want to `continue` with other variables, not `break` the `for` loop. Does this make sense to you?